motionplan/cBiRRT.go

```diff
@@ -110,115 +105,167 @@ func NewCBiRRTMotionPlannerWithSeed(frame referenceframe.Frame, nCPU int, seed *
 ) ([][]referenceframe.Input, error) {
 	if planOpts == nil {
-		planOpts = NewBasicPlannerOptions()
+		planOpts = newBasicPlannerOptions()
 	}
-	solutionChan := make(chan *planReturn, 1)
+	solutionChan := make(chan *rrtPlanReturn, 1)
 	utils.PanicCapturingGo(func() {
-		mp.planRunner(ctx, goal, seed, planOpts, nil, solutionChan)
+		mp.rrtBackgroundRunner(ctx, goal, seed, planOpts, initRRTMaps(), nil, solutionChan)
 	})
 	select {
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	case plan := <-solutionChan:
-		return plan.toInputs(), plan.err
+		return plan.ToInputs(), plan.err
 	}
 }
 
-// planRunner will execute the plan. When Plan() is called, it will call planRunner in a separate thread and wait for the results.
-// Separating this allows other things to call planRunner in parallel while also enabling the thread-agnostic Plan to be accessible.
-func (mp *cBiRRTMotionPlanner) planRunner(
+// rrtBackgroundRunner will execute the plan. Plan() will call RRTBackgroundRunner in a separate thread and wait for results.
+// Separating this allows other things to call RRTBackgroundRunner in parallel allowing the thread-agnostic Plan to be accessible.
+func (mp *cBiRRTMotionPlanner) rrtBackgroundRunner(
 	ctx context.Context,
 	goal spatialmath.Pose,
 	seed []referenceframe.Input,
-	planOpts *PlannerOptions,
+	planOpts *plannerOptions,
+	rm *rrtMaps,
 	endpointPreview chan node,
-	solutionChan chan *planReturn,
+	solutionChan chan *rrtPlanReturn,
 ) {
 	defer close(solutionChan)
 
 	// setup planner options
 	if planOpts == nil {
-		solutionChan <- &planReturn{err: errNoPlannerOptions}
+		solutionChan <- &rrtPlanReturn{err: errNoPlannerOptions}
 		return
 	}
 	algOpts, err := newCbirrtOptions(planOpts, mp.frame)
 	if err != nil {
-		solutionChan <- &planReturn{err: err}
+		solutionChan <- &rrtPlanReturn{err: err}
 		return
 	}
 
+	// initialize maps
+	corners := map[node]bool{}
+
+	// TODO(rb) package neighborManager better
+	nm := &neighborManager{nCPU: algOpts.Ncpu}
+	nmContext, cancel := context.WithCancel(ctx)
+	defer cancel()
+	mp.start = time.Now()
+
 	// get many potential end goals from IK solver
-	solutions, err := getSolutions(ctx, planOpts, mp.solver, goal, seed, mp.Frame())
-	if err != nil {
-		solutionChan <- &planReturn{err: err}
+	solutions, err := getSolutions(ctx, planOpts, mp.solver, goal, seed, mp.Frame(), mp.randseed.Int())
+
+	if err != nil && len(rm.goalMap) == 0 {
+		solutionChan <- &rrtPlanReturn{err: err}
 		return
 	}
 
-	// publish endpoint of plan if it is known
-	if planOpts.MaxSolutions == 1 && endpointPreview != nil {
-		endpointPreview <- solutions[0]
-		endpointPreview = nil
+	for i, solution := range solutions {
+		// Check if we can directly interpolate to any solutions
+		if i == 0 && mp.checkPath(planOpts, seed, solution.Q()) {
+			solutionChan <- &rrtPlanReturn{steps: []node{&basicNode{q: seed}, solution}}
+			return
+		}
+
+		// if we got more solutions, add them
+		if _, ok := rm.goalMap[solution]; !ok {
```

Comment:
[minor] I think we need to go through the logic of what to do if we have a rrt map passed in.  I think we will want to leverage things like its neighbor manager, solutions, corners, etc.  Maybe we can sync offline about this